lib/min_heap.rb

```diff
@@ -44,28 +50,51 @@ def to_s
   end
 
   # This method returns true if the heap is empty
-  # Time complexity: ?
-  # Space complexity: ?
+  # Time complexity: O(n)
+  # Space complexity: O(1)
   def empty?
-    raise NotImplementedError, "Method not implemented yet..."
+    return true if @store.length == 0
   end
 
   private
 
   # This helper method takes an index and
   #  moves it up the heap, if it is less than it's parent node.
   #  It could be **very** helpful for the add method.
-  # Time complexity: ?
-  # Space complexity: ?
+  # Time complexity: log(n)
+  # Space complexity: log(n)
   def heap_up(index)
-    
+    parent_index = (index - 1) / 2
+    return nil if @store.empty?
+    return if index == 0 || @store[index].key >= @store[parent_index].key
+
+    swap(index, parent_index)
+    return heap_up(parent_index)
   end
 
   # This helper method takes an index and 
   #  moves it up the heap if it's smaller
   #  than it's parent node.
   def heap_down(index)
-    raise NotImplementedError, "Method not implemented yet..."
+    lc_index = index * 2 + 1
+    rc_index = index * 2 + 2
+    return if @store[lc_index].nil? && @store[rc_index].nil?
+    return if index == @store.length - 1
```

Comment:
I think this line is unnecessary since if they are at the end of the array, they have no children.